modules/core/src/main/java/org/apache/ignite/internal/processors/cache/consistentcut/ConsistentCutManager.java

```diff
@@ -0,0 +1,481 @@
+            nearTxId.asIgniteUuid() + " " + txLastCutVer + " " + onePhaseCommit);
+
+        if (s.cutAwait().containsKey(nearTxId)) {
+            Long prepVer = s.cutAwait().remove(nearTxId);
+
+            if (prepVer != null) {
+                if (txLastCutVer >= s.ver() && txLastCutVer > prepVer)
+                    s.exclude(nearTxId);
+            }
+            else if (txLastCutVer >= s.ver())
+                s.exclude(nearTxId);
+
+            if (s.cutAwait().isEmpty() && s.complete()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Log ConsistentCutFinish(ver=" + s.ver() + ") on `handleRemoteTxCutVersion`" +
+                        " for tx " + nearTxId.asIgniteUuid());
+                }
+
+                logConsistentCutFinish(s.ver(), s.checkInclude());
+            }
+        }
+    }
+
+    /**
+     * @return CutVersion for specified transaction.
+     */
+    public long txCommitCutVer(IgniteTxAdapter tx) {
+        // Need lock here to avoid concurrent threads - that prepare FinishRequest and making ConsistentCut.
+        cutGuard.readLock().lock();
+
+        try {
+            ConsistentCutState s = lastCutState;
+
+            if (s == null)
+                return 0L;
+
+            if (tx.near() && !tx.onePhaseCommit()) {
+                if (!s.include().isEmpty() && s.include().contains(tx.nearXidVersion()))
+                    return s.ver() - 1;
+
+                if (!s.nearPrepare().isEmpty()) {
+                    Long v = s.nearPrepare().remove(tx.nearXidVersion());
+
+                    if (v != null)
+                        return v;
+                }
+            }
+            // Shows that this 1PC will be included to PREVIOUS cut (decrement lastCutVer to show that we need prev one).
+            else if (tx.dht() && !tx.local() && tx.onePhaseCommit() && s.include().contains(tx.nearXidVersion()))
+                return s.ver() - 1;
+
+            // Commit after Cut.
+            return s.ver();
+        }
+        finally {
+            cutGuard.readLock().unlock();
+        }
+
+    }
+
+    /**
+     * Handles a Consistent Cut Check event sent over discovery.
+     */
+    private void handleConsistentCutCheckEvent(ConsistentCutCheckDiscoveryMessage m) {
+        if (m.progress()) {
+            log.warning("Skip Consistent Cut procedure. Some nodes hasn't finished yet previous one.");
+
+            return;
+        }
+
+        if (skipConsistentCut())
+            m.inProgress();
+    }
+
+    /**
+     * Handles Consistent Cut Start event sent over discovery.
+     */
+    private void handleConsistentCutStartEvent() {
+        // Coordinator inits the Consistent Cut procedure on every node in a cluster (by Communication SPI).
+        // Then other nodes just skip the message.
+        if (!U.isLocalNodeCoordinator(cctx.discovery()))
+            return;
+
+        cctx.kernalContext().pools().consistentCutExecutorService().execute(() -> {
```

Comment:
1) Use system pool instead. Avoid additional ExecutorServices creation.
2) See no reason to perform this async, it's just a message resending.
3) As i mentioned some lines before, this shoild be simplified to 
- per node communication `ready` message to coordinator
- discovery message from coordinator when all `ready` message are received.